### ISSUE TITLE
Py3-Tracker

### ISSUE THREAD

**Issue by tstenner** [bleachbit/bleachbit#163]
Todo-tracker for supporting both Python 2 and 3:

- [X] print statements (PR #153)
- [x] require [six](https://pythonhosted.org/six/)
- [x] absolute imports
- [ ] unicode / str-split ([types for unit tests](http://stackoverflow.com/a/25935110))
- [x] merge GTK 3 port (PyGTK isn't available for Python 2)

**Comment by ROCKNROLLKID** [comment]
Wouldn't it be better to just have 2 branches, one with Python 2/GTK2 and then one with Python 3/GTK3, because if we are talking about wanting to shrink the installer and all, then this is probably going to have the opposite effect. The Python 2 branch can be just bug fixes for older systems and eventually drop support for it and move on to Python 3.

**Comment by tstenner** [comment]
There are some things (six / other compatibility libraries, GTK3) that might make a size difference, but most planned changes wouldn't. Maintaining two separate branches would be too much work (at least in my opinion), but I think it's too soon to drop Python 2 support.

**Comment by tstenner** [comment]
Debian stable has Python 3.4, RHEL [apparently too](https://dl.iuscommunity.org/pub/ius/stable/Redhat/6/x86_64/repoview/python34u.html).

**Comment by az0** [comment]
After the next stable release (and fixing any major regressions there) I would look at how much I can shrink Python 2 + GTK 3 on Windows.  Another option is considering GTK 2 and GTK 3 in the _same branch_ as separate GUIs provided through separate sets of files: this is how some applications support multiple front ends (e.g., Transmission BitTorrent client), but if GTK 3 can be shrunk to a reasonable size, I do not see much benefit in supporting GTK 2 much longer.

On the original check list above everything here seems like it would be fine to merge into the master any time soon.

**Comment by ROCKNROLLKID** [comment]
People can just use BleachBit 1.12 for older systems. Maybe Andrew can do a few bug fix patches for that, but ultimately there isn't really any serious issues with that version, other then maybe the previewing emptying recycling bin, which was fixed for 2.0. Maybe Andrew can do a hot fix patch to fix it for 1.12 so older systems can have a more stable BleachBit and include any new cleaners that were done recently.

I think something like this would be better because all of this is going to be a lot of work, only to have to do all the work taking it out, once Python 2 is done for. He hasn't merged all of Python 3, yet, so it isn't too late to do this.

**Comment by tstenner** [comment]
I'd also prefer to drop Python 2 support (makes it a lot easier), but it might be too early to do this.

**Comment by az0** [comment]
After the next stable release is out and significant regressions are fixed, then I will look more carefully at dropping Python 2 and moving to Python 3 and GTK 3. I will try not to delay this, and right now I am actively trying to  get to a beta release for BleachBit 2.0.

Tristan, perhaps you want to prioritize which Python 3 migration pieces you tackle based on how much of the effort will survive without rework.

**Comment by tstenner** [comment]
I'm also starting to like the idea of separate branches for GTK2+Py2 (Bleachbit Classic?) and GTK3+Py3.
After all, distros that don't have a recent version of Python 3 will still have PyGTK and probably won't need cleaners for many new programs.

Dropping support for Python 2 could also allow us to use type hints without comments or stub files and run the cleaners asynchronously (i.e. no more `if time > 10: yield True`) and so on.

**Comment by tstenner** [comment]
Of course, things might be different for the Windows version. I'm mostly using Python on Linux / Mac so I don't follow the state of Python on Windows.

**Comment by ROCKNROLLKID** [comment]
Python 3.4 can run on XP+ just fine. 3.5 dropped support for XP, though. Since you are targeting 3.4, as stated in the other PR, Windows shouldn't be a issue. Linux and Mac support is more of a concern for this new update and we have to figure out where we should draw the line when it comes to supporting older versions on Linux and Mac.

**Comment by tstenner** [comment]
We'd have to target 3.4 as PyGI doesn't support 3.5 yet.

**Comment by dotmagic** [comment]
@tstenner 

> Python 3.0 final was released on December 3rd, 2008

It's over 8 years ago! That's a incredible long time for software. So I can't agree with your earlier comment its too early to drop Python 2.x. I'd like to try the modernization branch next weekend.

**Comment by ROCKNROLLKID** [comment]
I think we came to the agreement already that if we are going to keep python/gtk 2, we should do it in separate branches, otherwise it is not worth doing.

**Comment by az0** [comment]
The last stable release of BleachBit was July 2016. There are many good changes nearly ready, and I do not want to postpone it longer with major changes.

I will consider a wholesale change to Python 3 and GTK 3 after BleachBit version 2.0 is published and regressions are stabilized, like with a follow-up bug fix release. I prefer not to maintain two branches.

**Comment by tstenner** [comment]
The str/byte-split is quite messy, especially with filesystems on all 3 platforms. I'd first ramp up the test coverage / polish the existing tests, wait until python 3.6 and [PEP 519](https://www.python.org/dev/peps/pep-0519/) is available and then use it.
This way we could have two major releases (bleachbit 2 with GTK 3 and bleachbit 3 with python 3) without too many breaking changes.

**Comment by az0** [comment]
Yes, the file system strings are complex. I was wondering whether there was a good way to rewrite that---like maybe with a special class that represents a path name with methods for file operations.

**Comment by tstenner** [comment]
As far as I understand it that's what PEP 519 is about, but I'd have to take a look at it first.

**Comment by hlef** [comment]
Debian is in the process of removing Python 2 from sid and bullseye. As of now, we are still building Bleachbit for Python 2 only. If this doesn't change soon, we will be forced to remove Bleachbit from the Debian archive (see our [bug report](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=936214) for more information). 

Bleachbit has a large user base in Debian, we would really like to avoid that!

I was wondering, how far are we from a final Python 3 release?

**Comment by az0** [comment]
A final release of BleachBit with Python 3 is at least a few months away. I am focusing hard to release the next BleachBit with GTK 3 support (after removing GTK 2). Then the focus will be Python 3 (completely replacing Python 3). Is there a deadline for Debian? If possible, I will try to manage to the deadline

**Comment by hlef** [comment]
> A final release of BleachBit with Python 3 is at least a few months away. I am focusing hard to release the next BleachBit with GTK 3 support (after removing GTK 2). Then the focus will be Python 3 (completely replacing Python 2). Is there a deadline for Debian? If possible, I will try to manage to the deadline

I am not aware of any "hard" deadline, but there was a heads-up recently and the migration work is quite active currently. I understand it's a lot of work for you, and will do my best to avoid bleachbit's removal until the release.

Thanks for your work!

**Comment by hlef** [comment]
@az0 thanks for your work on GTK3. Unfortunately Bleachbit has been marked for autoremoval in Debian testing as part of the Python 3 migration... It will be removed on 16.11.19.

Is there anything to do? Loosing Bleachbit would be a real pain for many Debian users.

**Comment by az0** [comment]
@hlef 
I don't know any way to finish by November 16, but I can make it a priority to do it soon. If I finished around, say, January 1, could BleachBit get added back in?

Will BleachBit remain for older versions of Debian that still use Python 2, or is the autoremoval retroactive?

**Comment by w-flo** [comment]
So I'm not a Debian developer or anything, but I'll try to explain the situation.

Older Debian versions are not affected by this, the package will be removed from Debian "testing" only. Debian "testing" is basically the development branch for the next Debian stable release (bullseye). If bleachbit is removed from testing, it will still be in Debian "unstable" and previous stable releases. Packages are automatically copied from unstable to testing if their dependencies are in testing and some other conditions are met (e.g. no release-critical bugs).

So if bleachbit's python2 dependency can be dropped in Debian unstable before the various freezes for the next stable release, and there are no other release-critical issues with the package at the time, it will be automatically re-added to testing. January 1 should be well before the freeze begins. Note that Ubuntu 20.04 LTS wants to remove python2 too, and their Debian import freeze is on February 27th, so the Debian package should no longer depend on Python 2 by then for Bleachbit to be included in the next Ubuntu LTS. (I believe they take packages from Debian unstable nowadays, but they used to pick them from Debian testing for LTS.)

Also, I get the impression that maybe filing the dependency on Python 2 as a "serious" bug (causing the auto-removal from testing countdown to start) was only intended for packages that still depend on pygtk(2). Bleachbit's dependency on pygtk in Debian testing was dropped on October 31st, and the bug was set to "serious" in a mass-raising of severities for multiple packages on November 1st. See the November 2nd message by Matthias Klose where he refers to pygtk a lot, so he probably didn't notice that pygtk is no longer a dependency: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=936214 Maybe someone should remove Bleachbit as a blocker for debian bug 937452.

He also mentions the possibility to lower the bug severity to cancel the auto-removal countdown, so maybe the auto-removal from testing can be avoided. Although there's an aggressive push to remove python2-dependent packages, and bleachbit is an easy target because nothing depends on it. Now that it no longer depends on old pygtk2 and there's a clear intention to port to python3, it may or may not be allowed to stay in testing for a bit longer.

**Comment by jbicha** [comment]
Yes, w-flo's analysis is good. I lowered the severity of the bleachbit bug in Debian so it won't be automatically removed from Bullseye in November. Your goal of January 1 sounds very reasonable to me. I don't expect there will be any disruption to any users of Debian or Debian derivatives if the porting is completed in January.

**Comment by az0** [comment]
OK, I will make a push for Python 3 with a goal of January 1. Thanks for the explainations.

**Comment by az0** [comment]
P.S., the first release of BleachBit on Python 3 might work only on Linux, not on Windows.

**Comment by az0** [comment]
Status update:
I just merged a pull request with significant progress into a new branch called py3. All the unit tests on Linux pass. Some of the CLI and GUI work. Windows doesn't work yet.

**Comment by sandrotosi** [comment]
hey @az0 we're heading towards the end of January, and a month past your last status update: how's the port going? asking for a friend (Debian) :)

**Comment by az0** [comment]
@sandrotosi 
Feel free to test it from the "py3" branch. File a new bug report for any issues. Linux works OK, but Windows could use more work. I'm considering a Linux-only release initially.

**Comment by az0** [comment]
By the way, there is a issue tracker label called [modernization](https://github.com/bleachbit/bleachbit/labels/modernization) to track issues related to Python 3. Today I recorded three new issues related to a new pull request. If you do early testing for Python 3, let's use this label.

**Comment by az0** [comment]
Python 3 is working rather well, so I am closing this. I'm going to check what's left to do a beta release
